### PR TITLE
Use jq instead of `deno eval` in CI [take 2]

### PR DESCRIPTION
**Pull request by lishaduck**
Try to reland #68

### CODE REVIEW

**Comment by lishaduck** [comment]
I'm not really sure how to test this. Should we make CI pipelines run on edit?

**Comment by lishaduck** [comment]
Oh, I see. It's running via pull_request_trigger, so it doesn't run my workflow for security reasons. I'd forgotten about that. I'll trigger the edge cases on my repo and report back.

**Comment by lishaduck** [comment]
Not quite. I've got a few more ideas up my sleeve.

**Comment by lishaduck** [comment]
Yeah! It works! ~~(← mostly)~~
`deno task dev` is running over the `dom` package, which doesn't have a deno.json. Is that intended?

**Comment by lishaduck** [comment]
> `deno task dev` is running over the `dom` package, which doesn't have a deno.json. Is that intended?

I see. It's not going to be triggered for that. Ok, it works then!
I also "explicitly" set pipefail, etc so that that won't be an issue in the future. (though bash is weird, and think it won't actually help with interpolation, per [oils](https://www.oilshell.org/release/latest/doc/error-handling.html#bash:~:text=%23%20Shell%20style-,x%3D%24(false),-sigpipe_status_ok%20Ignores%20an), which flagged this in some of my personal scripts a few weeks ago)

**Comment by lowlighter** [comment]
The dom package should be excluded now (I renamed it temporarly to `deno.json` instead of `deno.jsonc` to skip for now). 
It's indeed a WIP

Sorry I didn't see your PR I made some changes in the meantime to the CI by moving some stuff around to improve linting and testing.

If you have time, could you rebase your changes on main branch and reapply your changes ? (there's also one more deno.eval to patch in the linting job)

Btw is there any reason to explicitely add `shell: bash` to all jobs ? 
It kind of feel too verbose but if there's a catch with the implicit way, why not

**Comment by lishaduck** [comment]
> The dom package should be excluded now (I renamed it temporarly to `deno.json` instead of `deno.jsonc` to skip for now). It's indeed a WIP

Thanks for letting me know 😁 I was manually running it and it failed.

> Sorry I didn't see your PR I made some changes in the meantime to the CI by moving some stuff around to improve linting and testing.
>
> If you have time, could you rebase your changes on main branch and reapply your changes ? (there's also one more deno.eval to patch in the linting job)

That's fine, can do!
 
> Btw is there any reason to explicitely add `shell: bash` to all jobs ? It kind of feel too verbose but if there's a catch with the implicit way, why not

I agree it feels too verbose, but it does a few things:
1. It runs with `-o pipefail`, so that failing commands correctly fail the pipeline (main reason)
2. On Windows, it runs it on git bash, so that the syntax is still valid (good reason)
3. It makes it use bash on self-hosted runners that default to POSIX sh (and runs with `--noprofile --norc` so that it's fully reproducible, and faster) (ok reason)

`#3` doesn't apply to this, but if you decided to run tests on windows, `2` makes it trivial (at least, the CI part). `1` is definitely a win.

**Comment by lishaduck** [comment]
Ok, I think it should be good to go now!
Again, this is non-trivial to test, so I'd just merge it and revert it if it still doesn't work, but it's close enough I'd be willing to let it merge.

**Review by lowlighter — APPROVED** [review]

**Comment by lowlighter** [comment]
Actually looking at the GA docs, it seems it's possible to specify the default shell globally
https://docs.github.com/en/actions/using-jobs/setting-default-values-for-jobs#setting-default-shell-and-working-directory

It might be better to set it once per job rather than at every step

Seems like it's also possible to specifify the default `working-directory` per a job basis:
https://docs.github.com/en/actions/using-jobs/setting-default-values-for-jobs#example-setting-default-run-step-options-for-a-job
Could help for some cleanup too

**Comment by lishaduck** [comment]
> Actually looking at the GA docs, it seems it's possible to specify the default shell globally https://docs.github.com/en/actions/using-jobs/setting-default-values-for-jobs#setting-default-shell-and-working-directory
> 
> It might be better to set it once per job rather than at every step
> 
> Seems like it's also possible to specifify the default `working-directory` per a job basis: https://docs.github.com/en/actions/using-jobs/setting-default-values-for-jobs#example-setting-default-run-step-options-for-a-job Could help for some cleanup too

Ok, I can clean that up!

File: .github/workflows/ci_package.yml [hunk 1]
```diff
@@ -60,26 +65,28 @@ jobs:
...
+            .github/workflows/ci_package.yml
 
   lint:
     name: Lint
     if: (inputs.ref == 'refs/heads/main' && inputs.event == 'push') || (needs.setup.outputs.edited == 'true')
     runs-on: ubuntu-latest
     needs: setup
+    defaults:
+      run:
+        shell: bash
```
Comment (lowlighter) [inline .github/workflows/ci_package.yml]:
```suggestion
        working-directory: ${{ inputs.package }}
```

File: .github/workflows/ci_package.yml [hunk 2]
```diff
@@ -88,6 +95,10 @@ jobs:
     needs: setup
     permissions:
       security-events: write
+    defaults:
+      run:
+        shell: bash
```
Comment (lowlighter) [inline .github/workflows/ci_package.yml]: unecessary

File: .github/workflows/ci_package.yml [hunk 3]
```diff
@@ -111,6 +122,10 @@ jobs:
       slow_types: ${{ steps.package.outputs.slow_types }}
       tag: ${{ steps.tag.outputs.tag }}
       version: ${{ steps.tag.outputs.semver }}
+    defaults:
+      run:
+        shell: bash
```
Comment (lowlighter) [inline .github/workflows/ci_package.yml]:
```suggestion
        working-directory: ${{ inputs.package }}
```

File: .github/workflows/ci_package.yml [hunk 4]
```diff
@@ -169,6 +178,10 @@ jobs:
       id-token: write
       contents: write
       packages: write
+    defaults:
+      run:
+        shell: bash
```
Comment (lowlighter) [inline .github/workflows/ci_package.yml]:
```suggestion
        working-directory: ${{ inputs.package }}
```

**Review by lowlighter — CHANGES_REQUESTED** [review]

**Comment by lishaduck** [comment]
Oh, I guess my comment didn't get posted. I got some sleep, but I'm off to fix all those.

**Comment by lishaduck** [comment]
It works now, and is fully cleaned up! Thanks for your patience, it's been a while since I've JQed.
One thing I noticed is that you don't actually support saving comments in your config syncer, so it might be more unambiguous to use `deno.json` instead? Comments would break jq, though hopefully that'll change soon (jqlang/jq#2548 is held up on rebasing, etc, but is relatively likely to merge _someday_).

**Comment by lishaduck** [comment]
@lowlighter, ping :)
No need to merge if you're not on GH right now, but it's finished.

**Review by lowlighter — APPROVED** [review]

**Comment by lowlighter** [comment]
Seems to work great, thanks !

> One thing I noticed is that you don't actually support saving comments in your config syncer, so it might be more unambiguous to use `deno.json` instead? Comments would break jq, though hopefully that'll change soon ([jqlang/jq#2548](https://github.com/jqlang/jq/pull/2548) is held up on rebasing, etc, but is relatively likely to merge _someday_).

Yeah, but it's mostly because the JSONC library from deno/std doesn't support it, it's not really that I don't want to x)

**Comment by lishaduck** [comment]
> Seems to work great, thanks !

You're welcome!

> > One thing I noticed is that you don't actually support saving comments in your config syncer, so it might be more unambiguous to use `deno.json` instead? Comments would break jq, though hopefully that'll change soon ([jqlang/jq#2548](https://github.com/jqlang/jq/pull/2548) is held up on rebasing, etc, but is relatively likely to merge _someday_).
> 
> Yeah, but it's mostly because the JSONC library from deno/std doesn't support it, it's not really that I don't want to x)

Fair enough.

**Pull request merged**